dmoj/executors/base_executor.py

```diff
@@ -261,7 +270,53 @@ def get_env(self):
         return env
 
 
-class CompiledExecutor(BaseExecutor):
+# A lot of executors must do initialization during their constructors, which is
+# complicated by the CompiledExecutor compiling *during* its constructor. From a
+# user's perspective, though, once an Executor is instantiated, it should be ready
+# to launch (e.g. the user shouldn't have to care about compiling themselves). As
+# a compromise, we use a metaclass to compile after all constructors have ran.
+#
+# Using a metaclass also allows us to handle caching executors transparently.
+# Contract: if cached=True is specified and an entry exists in the cache,
+# `create_files` and `compile` will not be run, and `_executable` will be loaded
+# from the cache.
+class CompiledExecutorMeta(type):
+    def _cleanup_cache_entry(key, executor):
+        # Mark the executor as not-cached, so that if this is the very last reference
+        # to it, __del__ will clean it up.
+        executor.is_cached = False
+
+    compiled_binary_cache = pylru.lrucache(env.compiled_binary_cache_size, _cleanup_cache_entry)
+
+    def __call__(self, *args, **kwargs):
+        is_cached = kwargs.get('cached')
+        if is_cached:
+            kwargs['dest_dir'] = env.compiled_binary_cache_dir
+
+        # Finish running all constructors before compiling.
+        obj = super(CompiledExecutorMeta, self).__call__(*args, **kwargs)
+        obj.is_cached = is_cached
+
+        # Before writing sources to disk, check if we have this executor in our cache.
+        if is_cached:
+            cache_key = obj.get_binary_cache_key()
+            if cache_key in CompiledExecutorMeta.compiled_binary_cache:
+                executor = CompiledExecutorMeta.compiled_binary_cache[cache_key]
+                # Minimal sanity checking: is the file still there? If not, we'll just recompile.
+                if os.path.isfile(executor._executable):
+                    obj._executable = executor._executable
```

Comment:
I don't know how I feel about just setting the executable. This will make `self._dir` go out of sync with `self._executable`, as well as not having all the other files created by `create_files`.

Reply:
Alternatives:

* `deepcopy` the `executor` and return it instead (weird because constructor runs once and is then disposed)
* return the `executor` (weird for the same reason)
* copy over `_dir` as well

It's going to be an imperfect solution either way I think, the only question is which weird behaviour we prefer more.

Reply:
I think returning the old executor is what you'd expect from such caching, and would be the least disruptive. However, there are other arguments to the constructor that do not alter the generated executable.

The best solution would probably to copy over `_dir`.

Reply:
Done.